File: keyboards/android/app/src/main/java/com/lisanuddawat/keyboard/BiDiFixView.kt
package com.lisanuddawat.keyboard

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.os.Build
import android.view.View

// Animated button shown in the predictive bar when mixed-script text is detected.
// Renders a mini before→after of the specific BiDi issue: the LTR fragment slides
// from the wrong side to the correct side, then the animation loops.
class BiDiFixView(context: Context) : View(context) {

    private val rtlPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { textAlign = Paint.Align.LEFT }
    private val ltrPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { textAlign = Paint.Align.LEFT }

    private var rtlStr = "ع"
    private var ltrStr = "123"

    // 0 = broken state, 1 = fixed state
    private var animPhase = 0f

    private val animator = ValueAnimator.ofFloat(0f, 1f).apply {
        duration = 3200L
        repeatCount = ValueAnimator.INFINITE
        addUpdateListener { anim ->
            val r = anim.animatedValue as Float
            // 0–25 %: slide to fixed; 25–65 %: hold fixed; 65–80 %: slide back; 80–100 %: hold broken
            animPhase = when {
                r < 0.25f -> r / 0.25f
                r < 0.65f -> 1f
                r < 0.80f -> 1f - (r - 0.65f) / 0.15f
                else      -> 0f
            }
            invalidate()
        }
    }

    var onTap: (() -> Unit)? = null

    init {
        val size = dp(12f)
        rtlPaint.textSize = size
        ltrPaint.textSize = size * 0.9f

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            runCatching { resources.getFont(R.font.fatemi_maqala_regular) }.getOrNull()
                ?.let { rtlPaint.typeface = it }
        }

        isClickable = true
        isFocusable = true
        setOnClickListener { onTap?.invoke() }
        visibility = GONE
    }

    fun setIssue(issue: BiDiAnalyzer.Issue?) {
        if (issue != null) {
            rtlStr = issue.previewRtl
            ltrStr = issue.previewLtr
            if (visibility != VISIBLE) {
                visibility = VISIBLE
                animator.start()
            }
        } else {
            visibility = GONE
            animator.cancel()
            animPhase = 0f
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        animator.cancel()
    }

    override fun onDraw(canvas: Canvas) {
        val w = width.toFloat()
        val h = height.toFloat()

        val t = smoothStep(animPhase)
        updateColors(t)

        val rtlW    = rtlPaint.measureText(rtlStr)
        val ltrW    = ltrPaint.measureText(ltrStr)
        val gap     = dp(2.5f)
        val totalW  = rtlW + ltrW + gap
        val originX = (w - totalW) / 2f
        val baseline = h * 0.63f

        // Broken state: [ltr] [gap] [rtl]   →   Fixed: [rtl] [gap] [ltr]
        val ltrX = lerp(originX,             originX + rtlW + gap, t)
        val rtlX = lerp(originX + ltrW + gap, originX,             t)

        canvas.drawText(ltrStr, ltrX, baseline, ltrPaint)
        canvas.drawText(rtlStr, rtlX, baseline, rtlPaint)
    }

    private fun updateColors(t: Float) {
        val label  = KeyboardColors.keyLabel(context)
        val accent = Color.parseColor("#007AFF")
        val dim    = Color.argb(160, Color.red(label), Color.green(label), Color.blue(label))
        rtlPaint.color = lerpColor(label, accent, t)
        ltrPaint.color = lerpColor(dim,   accent, t)
    }

    private fun smoothStep(x: Float): Float {
        val t = x.coerceIn(0f, 1f)
        return t * t * (3f - 2f * t)
    }

    private fun lerp(a: Float, b: Float, t: Float) = a + (b - a) * t

    private fun lerpColor(from: Int, to: Int, t: Float): Int {
        fun ch(f: Int, s: Int) = (f + (s - f) * t).toInt().coerceIn(0, 255)
        return Color.argb(
            ch(Color.alpha(from), Color.alpha(to)),
            ch(Color.red(from),   Color.red(to)),
            ch(Color.green(from), Color.green(to)),
            ch(Color.blue(from),  Color.blue(to))
        )
    }

    private fun dp(v: Float) = v * resources.displayMetrics.density
}

File: keyboards/ios/LisanUdDawat/LSDLearningKB/BiDiFixView.swift
import UIKit

// Animated button shown in the predictive bar when mixed-script text is detected.
// Displays a mini before→after loop: the LTR fragment slides from the wrong
// side to the correct side and back, showing the user what the fix will do.

final class BiDiFixView: UIView {

    var onTap: (() -> Void)?

    private var rtlString = "ع"
    private var ltrString = "123"

    // 0 = broken state, 1 = fixed state
    private var animPhase: CGFloat = 0
    private var displayLink: CADisplayLink?
    private var startTime: CFTimeInterval = 0
    private let cycleDuration: CFTimeInterval = 3.2

    private let rtlLabel = UILabel()
    private let ltrLabel = UILabel()

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    required init?(coder: NSCoder) { fatalError() }

    private func setup() {
        rtlLabel.font = UIFont(name: "FatemiMaqala-Regular", size: 12) ?? .systemFont(ofSize: 12)
        ltrLabel.font = .systemFont(ofSize: 11)
        rtlLabel.textAlignment = .center
        ltrLabel.textAlignment = .center
        addSubview(rtlLabel)
        addSubview(ltrLabel)

        isUserInteractionEnabled = true
        addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(tapped)))
    }

    func setIssue(_ issue: BiDiAnalyzer.Issue?) {
        if let issue {
            rtlString = issue.previewRtl
            ltrString = issue.previewLtr
            if displayLink == nil { startAnimating() }
        } else {
            stopAnimating()
        }
    }

    // MARK: - Animation

    private func startAnimating() {
        startTime = CACurrentMediaTime()
        let dl = CADisplayLink(target: self, selector: #selector(tick))
        dl.add(to: .main, forMode: .common)
        displayLink = dl
    }

    private func stopAnimating() {
        displayLink?.invalidate()
        displayLink = nil
        animPhase = 0
        setNeedsLayout()
    }

    @objc private func tick() {
        let elapsed = CACurrentMediaTime() - startTime
        let r = Float(elapsed.truncatingRemainder(dividingBy: cycleDuration) / cycleDuration)
        let raw: Float
        switch r {
        case ..<0.25: raw = r / 0.25
        case ..<0.65: raw = 1.0
        case ..<0.80: raw = 1.0 - (r - 0.65) / 0.15
        default:      raw = 0.0
        }
        animPhase = CGFloat(smoothStep(raw))
        updateColors()
        setNeedsLayout()
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        let h = bounds.height
        let rtlW = rtlLabel.sizeThatFits(bounds.size).width + 2
        let ltrW = ltrLabel.sizeThatFits(bounds.size).width + 2
        let gap:  CGFloat = 3
        let total = rtlW + ltrW + gap
        let originX = (bounds.width - total) / 2

        // Broken state: [ltr][gap][rtl]  →  Fixed: [rtl][gap][ltr]
        let t = animPhase
        let ltrX = originX + (rtlW + gap) * t
        let rtlX = originX + (ltrW + gap) * (1 - t)

        ltrLabel.frame = CGRect(x: ltrX, y: 0, width: ltrW, height: h)
        rtlLabel.frame = CGRect(x: rtlX, y: 0, width: rtlW, height: h)

        rtlLabel.text = rtlString
        ltrLabel.text = ltrString
    }

    private func updateColors() {
        let accent = UIColor(red: 0, green: 0.478, blue: 1, alpha: 1)
        let base   = UIColor.label
        let dim    = UIColor.tertiaryLabel
        rtlLabel.textColor = interpolate(from: base,   to: accent, t: animPhase)
        ltrLabel.textColor = interpolate(from: dim,    to: accent, t: animPhase)
    }

    @objc private func tapped() { onTap?() }

    override func didMoveToWindow() {
        super.didMoveToWindow()
        if window == nil { stopAnimating() }
    }

    // MARK: - Helpers

    private func smoothStep(_ x: Float) -> Float {
        let t = max(0, min(1, x))
        return t * t * (3 - 2 * t)
    }

    private func interpolate(from a: UIColor, to b: UIColor, t: CGFloat) -> UIColor {
        var r1: CGFloat = 0; var g1: CGFloat = 0; var b1: CGFloat = 0; var a1: CGFloat = 0
        var r2: CGFloat = 0; var g2: CGFloat = 0; var b2: CGFloat = 0; var a2: CGFloat = 0
        a.getRed(&r1, green: &g1, blue: &b1, alpha: &a1)
        b.getRed(&r2, green: &g2, blue: &b2, alpha: &a2)
        return UIColor(red:   r1 + (r2 - r1) * t,
                       green: g1 + (g2 - g1) * t,
                       blue:  b1 + (b2 - b1) * t,
                       alpha: a1 + (a2 - a1) * t)
    }
}
